Frameworks/Foundation/NSScanner.mm

```diff
@@ -25,547 +25,311 @@
-                if (unicode == 'x' || unicode == 'X')
-                    break;
-
-            // fallthrough
-            case STATE_HEX:
-                if (unicode >= '0' && unicode <= '9') {
-                    if (!overflow) {
-                        unsigned check = (unsigned)(value * 16 + (unicode - '0'));
-                        if (check >= value)
-                            value = check;
-                        else {
-                            value = -1;
-                            overflow = YES;
-                        }
-                    }
-                } else if (unicode >= 'a' && unicode <= 'f') {
-                    if (!overflow) {
-                        unsigned check = (unsigned)(value * 16 + (unicode - 'a') + 10);
-                        if (check >= value)
-                            value = check;
-                        else {
-                            value = -1;
-                            overflow = YES;
-                        }
-                    }
-                } else if (unicode >= 'A' && unicode <= 'F') {
-                    if (!overflow) {
-                        unsigned check = (unsigned)(value * 16 + (unicode - 'A') + 10);
-
-                        if (check >= value)
-                            value = check;
-                        else {
-                            value = -1;
-                            overflow = YES;
-                        }
-                    }
-                } else {
-                    if (valuep) {
-                        *valuep = value;
-                    }
-
-                    return YES;
-                }
-                break;
-        }
-    }
-
-    if (hasValue) {
-        if (valuep) {
-            *valuep = value;
-        }
-
-        return YES;
-    }
-    return NO;
+    return __scanNumeric(self, valuep, _wcstoull_l, wcstoull, 16, true);
 }
 
 /**
  @Status Interoperable
 */
-- (BOOL)scanString:(id)string intoString:(NSString**)stringp {
-    unsigned int length = [_string length];
-    DWORD compareOption = 0;
-    NSRange range;
-
-    range.location = 0;
-    range.length = [string length];
-    int oldLocation = _location;
+- (BOOL)scanString:(NSString* _Nonnull)string intoString:(NSString* _Nullable*)stringp {
+    if (string.length == 0) {
+        return NO;
+    }
 
-    BOOL result = NO;
+    NSStringCompareOptions compareOption = 0;
+    NSRange range = { 0, string.length };
 
-    if (!_isCaseSensitive) {
-        compareOption = 1; // NSCaseInsensitiveSearch
+    if (!_caseSensitive) {
+        compareOption = NSCaseInsensitiveSearch;
     }
 
-    for (; _location < length; _location++) {
-        unichar unicode = (unichar)[_string characterAtIndex:_location];
-        range.location = _location;
-        if ((_len - _location) < [string length]) {
-            result = NO;
-            break;
+    const NSUInteger length = _string.length;
+    NSUInteger i = [self _indexOfNextUnskippedCharacter];
```

Comment:
it could in this instance, but there are other instances in the file where i needs to be available outside the for loop and I'd rather keep them in line